torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1362,160 @@ def rbinary_magic_impl(self, other):
+                    denominator_list.append(sympy.Pow(item.base, -item.exp))
+                else:
+                    numerator_list.append(item)
+
+            numerator = self._int(1) if len(numerator_list) == 0 else math.prod(self(numerator_list))
+
+            if len(denominator_list) > 0:
+                denominator = math.prod(self(denominator_list))
+                return self._div(numerator, denominator)
+
+            return numerator
+
+        def _Pow(self, expr: sympy.Pow):
+            base, exp = self(expr.args)
+            return base ** exp
+
+        def _FloorDiv(self, expr: FloorDiv):
+            base, divisor = self(expr.args)
+            return base / divisor
+
+        def __call__(self, expr):
+            if isinstance(expr, (tuple, list)):
+                return type(expr)(self(e) for e in expr)
+            if isinstance(expr, sympy.Rel):
+                typename = "Rel"
+            elif isinstance(expr, sympy.Integer):
+                typename = "int"
+                expr = int(expr)
+            elif isinstance(expr, sympy.core.numbers.IntegerConstant):
+                constantmap = {
+                    sympy.S.Zero: 0,
+                    sympy.S.One: 1,
+                    sympy.S.NegativeOne: -1,
+                }
+                typename = "int"
+                expr = constantmap[expr]
+            else:
+                typename = type(expr).__name__
+
+            method = f"_{typename}"
+            assert hasattr(self, method), (
+                f"unsupported sympy operation: {expr} ({method})"
+            )
+            return getattr(self, method)(expr)
+
+    class TranslationValidator:
+        def __init__(self) -> None:
+            super().__init__()
+            self._symbols = {}
+            self._inputs = set()
+            self._outputs = set()
+            self._z3 = SympyToZ3(self._symbols)
+
+        def _add_freesymbols(self, e: sympy.Expr) -> None:
+            for s in e.free_symbols:
+                if s not in self._symbols:
+                    assert isinstance(s, sympy.Symbol) and s.is_integer and s.is_positive
+                    self._symbols[s] = z3.Int(s.name)
+                    self._outputs.add(self._symbols[s] >= 0)
+
+        def _sympy_to_z3(self, expr: sympy.Expr):
+            z3_expr = self._z3(expr)
+            return z3_expr
+
+        def add_input(self, e: sympy.Expr) -> None:
+            self._add_freesymbols(e)
+            self._inputs.add(self._sympy_to_z3(e))
+
+        def add_output(self, e: sympy.Expr) -> None:
+            self._add_freesymbols(e)
+            self._outputs.add(self._sympy_to_z3(e))
+
+        def validate(self) -> bool:
+            from torch._dynamo.utils import dynamo_timed
+
+            if len(self._inputs) == 0 or len(self._outputs) == 0:
+                return True
+
+            solver = z3.Solver()
```

Comment:
the SimpleSolver() should be faster

Reply:
In fact, I was using `SolveFor('QF_NRA')`, here. Do you think I should stick to `SimpleSolver`?

Reply:
NRA is for reals. Now you stretch my knowledge of this part of Z3; I don't now if the NRA solver supports Ints or not.
I guess it's a matter of benchmarking which performs better: QF_NRA or SimpleSolver.